python/JSONtoObj.c

```diff
@@ -185,40 +187,60 @@ PyObject* JSONToObj(PyObject* self, PyObject *args, PyObject *kwargs)
       return NULL;
   }
 
-  if (PyBytes_Check(arg))
+  Py_buffer buffer;
+  size_t sarg_length;
+  char * raw;
+  bool is_bytes_like = !PyObject_GetBuffer(arg, &buffer, PyBUF_C_CONTIGUOUS);
+  if (is_bytes_like)
   {
-      sarg = arg;
+    #ifdef PYPY_VERSION
+      // PyPy's buffer protocol implementation is buggy: https://foss.heptapod.net/pypy/pypy/-/issues/3872
+      if (!PyBytes_Check(arg) && !PyByteArray_Check(arg)) {
+        PyBuffer_Release(&buffer);
+        PyErr_Format(PyExc_TypeError, "Arbitrary bytes-like objects are not supported on PyPy, Use either string, bytes, or bytearray");
+        return NULL;
+      }
+    #endif
+    raw = buffer.buf;
+    sarg_length = buffer.len;
   }
-  else
-  if (PyUnicode_Check(arg))
-  {
-    sarg = PyUnicode_AsEncodedString(arg, NULL, "surrogatepass");
-    if (sarg == NULL)
+  else {
+    PyErr_Clear();
+    if (PyUnicode_Check(arg))
+    {
+      sarg = PyUnicode_AsEncodedString(arg, NULL, "surrogatepass");
+      if (sarg == NULL)
+      {
+        //Exception raised above us by codec according to docs
+        return NULL;
+      }
+      sarg_length = PyBytes_Size(sarg);
+      raw = PyBytes_AsString(sarg);
+    }
+    else
     {
-      //Exception raised above us by codec according to docs
+      PyErr_Format(PyExc_TypeError, "Expected String or C contiguous Bytes-like object");
```

Comment:
The capitalisation (which I know predates this change) is odd. But more importantly, I think we should add another PyPy check here so the error messages aren't confusing under PyPy when passing a bytes-like object that isn't `bytes` or `bytearray` vs passing something completely incompatible like an `int`.

```suggestion
      #ifdef PYPY_VERSION
        PyErr_Format(PyExc_TypeError, "Expected string, bytes, or bytearray");
      #else
        PyErr_Format(PyExc_TypeError, "Expected string or C-contiguous bytes-like object");
      #endif
```